### ISSUE TITLE
let_and_return false positive

### ISSUE THREAD

**Issue by leonardo-m** [Manishearth/rust-clippy#1669]
```
#![feature(conservative_impl_trait)]
fn pairs<'a, T>(items: &'a [T]) -> impl Iterator<Item=(&'a T, &'a T)> + 'a {
    items.iter().map(|i| (i, i))
}
fn foo() -> usize {
    let data = vec![0, 1, 2, 3];
    let result = pairs(&data).count();
    result
}
fn main() {
    foo();
}
```

```
warning: returning the result of a let binding from a block. Consider returning the expression directly.
 --> src\main.rs:8:5
  |
8 |     result
  |     ^^^^^^
  |
  = note: #[warn(let_and_return)] on by default
note: this expression can be directly returned
 --> src\main.rs:7:18
  |
7 |     let result = pairs(&data).count();
  |                  ^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#let_and_return
```

If I modify the code as suggested, it stops compiling:

```
#![feature(conservative_impl_trait)]
fn pairs<'a, T>(items: &'a [T]) -> impl Iterator<Item=(&'a T, &'a T)> + 'a {
    items.iter().map(|i| (i, i))
}
fn foo() -> usize {
    let data = vec![0, 1, 2, 3];
    pairs(&data).count()
}
fn main() {
    foo();
}
```

Rustc gives:

```
error: `data` does not live long enough
 --> ...\main.rs:8:1
  |
7 |     pairs(&data).count()
  |            ---- borrow occurs here
8 | }
  | ^ `data` dropped here while still borrowed
  |
  = note: values in a scope are dropped in the opposite order they are created
```

**Comment by ebroto** [comment]
Duplicate of #1524 (I missed it yesterday), also builds on stable ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1d89762b6f897677ec59897cd19717c))
